pipeline/src/main/java/io/datacater/Pipeline.java

```diff
@@ -0,0 +1,99 @@
+package io.datacater;
+
+import io.datacater.exceptions.TransformationException;
+import io.smallrye.common.annotation.Blocking;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.smallrye.reactive.messaging.kafka.Record;
+
+import java.util.UUID;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import io.vertx.mutiny.core.Vertx;
+import io.vertx.mutiny.core.buffer.Buffer;
+import io.vertx.mutiny.ext.web.client.HttpRequest;
+import io.vertx.mutiny.ext.web.client.HttpResponse;
+import io.vertx.mutiny.ext.web.client.WebClient;
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.eclipse.microprofile.reactive.messaging.Acknowledgment;
+import org.eclipse.microprofile.reactive.messaging.Incoming;
+import org.eclipse.microprofile.reactive.messaging.Outgoing;
+import org.jboss.logging.Logger;
+import org.jboss.resteasy.reactive.RestResponse;
+
+@ApplicationScoped
+public class Pipeline {
+  private static final Logger LOGGER = Logger.getLogger(Pipeline.class);
+
+  private final Integer DATACATER_PYTHONRUNNER_PORT =
+          ConfigProvider.getConfig()
+                  .getOptionalValue("datacater.python-runner.port", Integer.class)
+                  .orElse(50000);
+
+  private final String DATACATER_PYTHONRUNNER_HOST =
+          ConfigProvider.getConfig()
+                  .getOptionalValue("datacater.python-runner.host", String.class)
+                  .orElse("localhost");
+
+  private String host;
+  private Integer port;
+
+  private static final String PIPELINE_IN = "pipeline-in";
+  private static final String PIPELINE_OUT = "pipeline-out";
+  private static final String PIPELINE_ERROR_MSG = "Pipeline could not process message.\n Key: %s\n Value: %s";
+
+  WebClient client;
+
+  @Inject
+  void VertxResource(Vertx vertx) {
+    this.client = WebClient.create(vertx);
+  }
+
+  @Incoming(PIPELINE_IN)
+  @Outgoing(PIPELINE_OUT)
+  @Blocking
+  @Acknowledgment(Acknowledgment.Strategy.POST_PROCESSING)
+  public Record<UUID, JsonObject> processUUID(Record<UUID, JsonObject> message) {
+    // This is deliberately blocking
+    JsonObject transformedMessage = handleMessage(message);
```

Comment:
At the moment, the pipeline processes record by record.

Could we turn this into micro-batching records for performance reasons? An user could still use a batch size of 1 if they would prefer the record-by-record processing.

Reply:
If the record gets "filtered out", the Python runner returns `null`.
We should reflect this here.

Reply:
> At the moment, the pipeline processes record by record.
> 
> Could we turn this into micro-batching records for performance reasons? An user could still use a batch size of 1 if they would prefer the record-by-record processing.

I looked into this and wrote down some notes on how the processing would look to batch process a stream:

- They might lose their order going back and forth to python runner - still need to test this - probably need to manually enforce order based on offset
- also, if there is an exception for one record and we want to nack it, the entire batch is nacked.
- Depending on acknowledgment strategy (i.e. processing continues on nack), this could be fatal
- The processing from our side requires us to transform the messages to and from lists multiple times
    - we receive as a list<UUID, JsonObject>
    - we need to forward to python-runner as jsonarray<jsonobject> which requires us to iterate over list and transform inside pipeline
    - we receive jsonobject from python runner containing multiple nodes which each hold a message
    - we need to check the metadata of each message received to filter and log failed processing
    - we need to transform the jsonobject back into list<UUID, JsonObject> and probably sort by offset, so that 
    - the messages are back in the original order.

Do we think the python-runner performance boost is high enough to even out the extra processing done in pipeline?

Reply:
> I looked into this and wrote down some notes on how the processing would look to batch process a stream:

Thanks a lot. That is very helpful! :)
 
>     * They might lose their order going back and forth to python runner - still need to test this - probably need to manually enforce order based on offset

I feel that this is fine because records end up in a different topic/partition, where we cannot guarantee any particular order.

>     * also, if there is an exception for one record and we want to nack it, the entire batch is backed.

True! That is a downside. We must ack all records and handle exceptions inside the pipeline, e.g., by logging the exception/error, by writing the affected records to a dead-letter queue, or any other mean. 

>     * Depending on acknowledgment strategy (i.e. processing continues on nack), this could be fatal

Agreed. We should start with a simple acknowledgment strategy and extend to other strategies in the future.

>     * The processing from our side requires us to transform the messages to and from lists multiple times
>       
>       * we receive as a list<UUID, JsonObject>
>       * we need to forward to python-runner as jsonarray which requires us to iterate over list and transform inside pipeline
>       * we receive jsonobject from python runner containing multiple nodes which each hold a message
>       * we need to check the metadata of each message received to filter and log failed processing
>       * we need to transform the jsonobject back into list<UUID, JsonObject> and probably sort by offset, so that
>       * the messages are back in the original order.
> 
> 
> Do we think the python-runner performance boost is high enough to even out the extra processing done in pipeline?

Yes, this should still give us a huge performance boost. 🚀  Also, from my gut feeling, handling one JSON array containing X JSON objects should not take much more time than handling X single JSON objects, if we are looking at throughput. We might want to test it though to have more confidence.

Reply:
Re 1: Makes sense, thanks!
Re 2: I believe this is a bug, which has been fixed with c7cfbac1cfccd131d34a5fb93a18696dd7f40330 (that is not yet merged into main).